### ISSUE TITLE
BUG: weekday property raises TypeError in groupby method

### ISSUE THREAD

**Issue by philipphanemann** [pandas-dev/pandas#33752]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
n = 1000
s = pd.Series(range(n), index=pd.date_range('2020-01-01', periods=n, freq='H'))
s.groupby(lambda x: (x.weekday, x.hour)).mean()
```

#### Problem description
Used in the groupby method as a function argument in combination with  another index time attribute `weekday` raises a TypeError whereas `dayofweek` works although they are both aliases for each other. 

#### Expected Output
I do expect the same output as with `
s.groupby(lambda x: (x.dayofweek, x.hour)).mean()`

since it works as a single argument:

 `s.groupby(lambda x: x.dayofweek).mean().equals(s.groupby(lambda x: x.weekday).mean())`

#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.2.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-96-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : de_DE.UTF-8
LOCALE           : de_DE.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

**Comment by CloseChoice** [comment]
Can reproduce the bug on master. The difference is, that in `s.groupby(lambda x: (x.dayofweek, x.hour)).mean()` the values are internally interpreteded as python integeres while they are 
`<built-in method weekday of Timestamp object at 0x7efbdce974c0>`
 in 
`s.groupby(lambda x: (x.weekday, x.hour)).mean()`. And `<` is not implemented for these types. Is this an issue with the Cython class?

**Comment by rhshadrach** [comment]
Unlike `hour`, `weekday` is a method. I believe you get the correct result with:

````
import pandas as pd
n = 1000
s = pd.Series(range(n), index=pd.date_range('2020-01-01', periods=n, freq='H'))
s.groupby(lambda x: (x.weekday(), x.hour)).mean()
````

which gives the result:

````
(0, 0)     540
(0, 1)     541
(0, 2)     542
(0, 3)     543
(0, 4)     544
          ... 
(6, 19)    535
(6, 20)    536
(6, 21)    537
(6, 22)    538
(6, 23)    539
Length: 168, dtype: int64
````

Please let me know if I've missed something and I'll reopen.

**Comment by philipphanemann** [comment]
You are right. But what I now find confusing about it is, that `weekday`is a property, if applied on the index. 

i.e. `s.index.weekday`

but if I want to apply it on an index instance, it is a method. 

e.g. `s.index[0].weekday()`

with `dayofweek` this inconsistency does not exist. In both cases, it is used as a property. I don't know whether this might be the correct thread to talk about such issues?

**Comment by rhshadrach** [comment]
Thanks for responding - I think you make a good point. It seems to me we can add the following to the `_Timestamp` class in timestamps.pyx:

```
    @property
    def weekday(self) -> int:
        """
        Return day of the week.
        """
        return super().weekday()
```

Any thoughts here @jbrockmendel?

**Comment by jbrockmendel** [comment]
I think the Timestamp behavior is "right" here, and if anything we would change the other behavior to match it (with a deprecation cycle if necessary)

**Comment by rhshadrach** [comment]
I'm guessing the desire is to align with python's datetime. There, weekday is a method rather than a property, and that's where the consistency should lie. Is that right?

**Comment by jbrockmendel** [comment]
> I'm guessing the desire is to align with python's datetime. There, weekday is a method rather than a property, and that's where the consistency should lie. Is that right?

Correct.
